src/bidiMapper/domains/script/channelProxy.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  static async init(
+    channel: Script.ChannelProperties,
+    eventManager: IEventManager,
+    realm: Realm
+  ): Promise<ChannelProxy> {
+    const channelHandle = await ChannelProxy.#createHandle(realm);
+    const sendMessageHandle = await ChannelProxy.#createSendMessageHandle(
+      realm,
+      channelHandle
+    );
+
+    const channelProxy = new ChannelProxy(
+      channel,
+      eventManager,
+      realm,
+      channelHandle,
+      sendMessageHandle
+    );
+
+    void channelProxy.initChannelListener();
+    return channelProxy;
+  }
+
+  static async #createHandle(realm: Realm): Promise<CommonDataTypes.Handle> {
+    const createChannelHandleResult = await realm.cdpClient.sendCommand(
+      'Runtime.callFunctionOn',
+      {
+        functionDeclaration: String(() => {
+          const queue: unknown[] = [];
+          let queueNonEmptyResolver: null | (() => void) = null;
+
+          return {
+            /**
+             * Gets a promise, which is resolved as soon as a message occurs
+             * in the queue.
+             */
+            async getMessage(): Promise<unknown> {
+              const onMessage =
+                queue.length > 0
+                  ? Promise.resolve()
+                  : new Promise<void>((resolve) => {
+                      queueNonEmptyResolver = resolve;
+                    });
+              await onMessage;
+              return queue.shift();
+            },
+
+            /**
+             * Adds a message to the queue.
+             * Resolves the pending promise if needed.
+             */
+            sendMessage(message: string) {
+              queue.push(message);
+              if (queueNonEmptyResolver !== null) {
+                queueNonEmptyResolver();
+                queueNonEmptyResolver = null;
+              }
+            },
+          };
+        }),
+        executionContextId: realm.executionContextId,
+        serializationOptions: {
+          serialization: 'idOnly',
+        },
+      }
+    );
+    if (createChannelHandleResult.exceptionDetails) {
+      throw new Error(
+        `Failed to create channel handle: ${createChannelHandleResult.exceptionDetails}`
+      );
+    }
+    return createChannelHandleResult.result.objectId!;
+  }
+
+  static async #createSendMessageHandle(
+    realm: Realm,
+    channelHandle: CommonDataTypes.Handle
+  ): Promise<Handle> {
+    const sendMessageArgResult = await realm.cdpClient.sendCommand(
+      'Runtime.callFunctionOn',
+      {
+        functionDeclaration: String(
+          (channelHandle: {sendMessage: (message: string) => void}) => {
+            return channelHandle.sendMessage;
+          }
+        ),
+        arguments: [{objectId: channelHandle}],
+        executionContextId: realm.executionContextId,
+        serializationOptions: {
+          serialization: 'deep',
+        },
+      }
+    );
+    return sendMessageArgResult.result.objectId!;
```

Comment:
I would inline these (i.e. delete the `const` declaration)

Reply:
IMO it would be way less readable then

Reply:
OK